src/Microsoft.Identity.Client/Platforms/iOS/iOSBroker.cs

```diff
@@ -0,0 +1,236 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using Microsoft.Identity.Client.Core;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Utils;
+using UIKit;
+using Foundation;
+using System;
+using CoreFoundation;
+using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.Internal.Broker;
+using Microsoft.Identity.Client.UI;
+using Microsoft.Identity.Client.AppConfig;
+using Microsoft.Identity.Client.Exceptions;
+using Microsoft.Identity.Client.ApiConfig;
+
+namespace Microsoft.Identity.Client.Platforms.iOS
+{
+    /// <summary>
+    /// Handles requests which invoke the broker. This is only for mobile (iOS and Android) scenarios.
+    /// </summary>
+    internal class iOSBroker : NSObject, IBroker
+    {
+        private static SemaphoreSlim brokerResponseReady = null;
+        private static NSUrl brokerResponse = null;
+
+        private Dictionary<string, string> _brokerPayload;
+
+        private readonly ICoreLogger _logger;
+        private IServiceBundle _serviceBundle;
+
+        public iOSBroker(ICoreLogger logger)
+        {
+            _logger = logger;
+        }
+
+        public bool CanInvokeBroker(OwnerUiParent uiParent, IServiceBundle serviceBundle)
+        {
+            _serviceBundle = serviceBundle;
+
+            if (uiParent == null)
+            {
+                _logger.Verbose(iOSBrokerConstants.UiParentIsNullCannotInvokeBroker);
+                return false;
+            }
+
+            if (uiParent.CoreUiParent.CallerViewController == null)
+            {
+                _logger.Verbose(iOSBrokerConstants.CallerViewControllerIsNullCannotInvokeBroker);
+                return false;
+            }
+
+            var result = false;
+
+            if (_serviceBundle.Config.IsBrokerEnabled)
+            {
+                _logger.Verbose(iOSBrokerConstants.CanInvokeBroker + _serviceBundle.Config.IsBrokerEnabled);
+
+                uiParent.CoreUiParent.CallerViewController.InvokeOnMainThread(() =>
+                {
+                    result = UIApplication.SharedApplication.CanOpenUrl(new NSUrl("msauthv2://"));
+                });
+            }
+            if (!result)
+            {
+                _logger.Verbose(result + iOSBrokerConstants.CanInvokeBrokerReturnsFalseMessage);
+            }
+
+            return result;
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenUsingBrokerAsync(Dictionary<string, string> brokerPayload, IServiceBundle serviceBundle)
+        {
+            _brokerPayload = brokerPayload;
+
+            CheckBrokerPayloadForSilentFlow();
```

Comment:
why makes this check and fail? the method signature doesn't suggest this is a silent call

Reply:
we have something similar in adal. but if we control the broker payload, then we probably don't need this check here and it can be removed.

---
In reply to: [256626927](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/881#discussion_r256626927) [](ancestors = 256626927)